src/storage.rs

```diff
@@ -14,99 +16,334 @@ pub trait StorageBase {
-            _ => None
-        })
+        self.map.0.get_mut(&(entity.get_id() as u32))
+            .and_then(|x| if x.generation == entity.get_gen() { Some(&mut x.data) } else { None })
+
     }
     fn insert(&mut self, entity: Entity, value: T) {
-        while self.0.len() <= entity.get_id() {
-            self.0.push(None);
+        let value = GenerationData{
+            data: value,
+            generation: entity.get_gen()
+        };
+        if self.map.0.insert(entity.get_id() as u32, value).is_none() {
+            self.set.add(entity.get_id() as u32);
         }
-        self.0[entity.get_id()] = Some((entity.get_gen(), value));
     }
     fn remove(&mut self, entity: Entity) -> Option<T> {
-        self.0.get_mut(entity.get_id()).and_then(|x| {
-            if let &mut Some((gen, _)) = x {
-                // if the generation does not match avoid deleting
-                if gen != entity.get_gen() {
-                    return None;
-                }
+        if self.set.remove(entity.get_id() as u32) {
+            let value = self.map.0.remove(&(entity.get_id() as u32)).unwrap();
+            if value.generation == entity.get_gen() {
+                Some(value.data)
+            } else {
+                // it should be unlikely that the generation mismatch was
+                // wrong, so this is a slow path to re-add the value back
+                self.insert(entity, value.data);
+                None
             }
-            x.take().map(|(_, x)| x)
-        })
+        } else {
+            None
+        }
+    }
+    fn open(&self) -> (&BitSet, &Self::UnprotectedStorage) {
+        (&self.set, &self.map)
+    }
+    fn open_mut(&mut self) -> (&BitSet, &mut Self::UnprotectedStorage) {
+        (&self.set, &mut self.map)
     }
 }
 
-/// HashMap-based storage. Best suited for rare components.
-#[derive(Debug)]
-pub struct HashMapStorage<T>(pub HashMap<Entity, T, BuildHasherDefault<FnvHasher>>);
+impl<T> UnprotectedStorage<T> for InnerHashMap<T> {
+    unsafe fn get(&self, e: Index) -> &T {
+        &self.0.get(&e).unwrap().data
+    }
+    unsafe fn get_mut(&mut self, e: Index) -> &mut T {
+        &mut self.0.get_mut(&e).unwrap().data
+    }
+}
 
-impl<T> StorageBase for HashMapStorage<T> {
-    fn del(&mut self, entity: Entity) {
-        self.0.remove(&entity);
+pub struct InnerVec<T>(Vec<GenerationData<T>>);
+
+pub struct GenerationData<T> {
+    pub generation: Generation,
+    pub data: T
+}
+
+
+/// Vec-based storage, actually wraps data into options and stores the generations
+/// of the data in order to match with given entities. Supposed to have maximum
+/// performance for the components mostly present in entities.
+pub struct VecStorage<T> {
+    set: BitSet,
+    values: InnerVec<T>,
+}
+
+impl<T> VecStorage<T> {
+    fn extend(&mut self, id: usize) {
+        unsafe {
+            let delta = (id + 1) - self.values.0.len();
```

Comment:
also, the `let delta` line as well as the `debug_assert!` should be moved up, outside of `unsafe{}`